File: Practice/Graphs/Find if Path Exists in Graph/Find if Path Exists in Graph.py
from collections import defaultdict


class Solution:
    def traverseNodes(self, n, adjacency_list, source, destination, visited_nodes):
        # base case
        if source == destination:
            return True

        visited_nodes[source] = True

        # using adjacency_matrix
        # traversing the neighbours => since matrix we need to check all n and then check if 1
        # for i in range(n):
        #     if adjacency_matrix[source][i] and not visited_nodes[i]:
        #         if self.traverseNodes(n, adjacency_matrix, i, destination, visited_nodes):
        #             return True

        for neighbour in adjacency_list[source]:
            if not visited_nodes[neighbour]:
                # need to check if that neighboure can reach the destination => if so we will simply return
                if self.traverseNodes(
                    n, adjacency_list, neighbour, destination, visited_nodes
                ):
                    return True

        return False

    def validPath(
        self, n: int, edges: List[List[int]], source: int, destination: int
    ) -> bool:
        # using adjacency matrix
        # adjacency_matrix = [[0] * n for _ in range(n)]

        adjacency_list = defaultdict(list)
        visited_nodes = [False] * n

        for edge in edges:
            u, v = edge[0], edge[1]
            adjacency_list[u].append(v)
            adjacency_list[v].append(u)

        return self.traverseNodes(n, adjacency_list, source, destination, visited_nodes)

File: Practice/Graphs/Find if Path Exists in Graph/Find if Path Exists in Graph.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private boolean dfs(List<List<Integer>> adjacencyList, int source, int destination, int[] visited) {
        if (source == destination)
            return true;
        visited[source] = 1;

        for (int neighbor : adjacencyList.get(source)) {
            if (visited[neighbor] == 0 && dfs(adjacencyList, neighbor, destination, visited))
                return true;
        }

        return false;
    }

    public boolean validPath(int n, int[][] edges, int source, int destination) {
        List<List<Integer>> adjacencyList = new ArrayList<>();
        int[] visited = new int[n];

        for (int i = 0; i < n; i++) {
            adjacencyList.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];

            adjacencyList.get(u).add(v);
            adjacencyList.get(v).add(u);
        }

        return dfs(adjacencyList, source, destination, visited);
    }
}